src/main/java/io/jenkins/plugins/postgresql/PostgreSQLFingerprintStorage.java

```diff
@@ -0,0 +1,365 @@
+                List<String> facetEntries = entry.getValue();
+
+                for (String facetEntry : facetEntries) {
+                    try (PreparedStatement preparedStatement = connection.prepareStatement(
+                            Queries.getQuery("insert_fingerprint_facet_relation"))) {
+                        preparedStatement.setString(1, fingerprint.getHashString());
+                        preparedStatement.setString(2, instanceId);
+                        preparedStatement.setString(3, facetName);
+                        preparedStatement.setString(4, facetEntry);
+
+                        preparedStatement.executeUpdate();
+                    }
+                }
+            }
+            connection.commit();
+        } catch (SQLException e) {
+            LOGGER.log(Level.WARNING, "PostgreSQL failed in saving fingerprint: " + fingerprint.toString(), e);
+            throw new IOException(e);
+        }
+    }
+
+    /**
+     * Returns the fingerprint associated with the given unique id and the Jenkins instance ID, from the PostgreSQL
+     * instance.
+     */
+    public @CheckForNull Fingerprint load(@NonNull String id) throws IOException {
+        Fingerprint fingerprint = null;
+        try (Connection connection = getConnection()) {
+            Map<String,String> fingerprintMetadata;
+
+            try (PreparedStatement preparedStatement = connection.prepareStatement(
+                    Queries.getQuery("select_fingerprint"))) {
+                preparedStatement.setString(1, id);
+                preparedStatement.setString(2, instanceId);
+                ResultSet resultSet = preparedStatement.executeQuery();
+                fingerprintMetadata = DataConversion.extractFingerprintMetadata(resultSet, id);
+            }
+
+            if (fingerprintMetadata.size() == 0) {
+                return null;
+            }
+
+            Map<String, Fingerprint.RangeSet> usageMetadata;
+
+            try (PreparedStatement preparedStatement = connection.prepareStatement(
+                    Queries.getQuery("select_fingerprint_job_build_relation"))) {
+                preparedStatement.setString(1, id);
+                preparedStatement.setString(2, instanceId);
+                ResultSet resultSet = preparedStatement.executeQuery();
+                usageMetadata = DataConversion.extractUsageMetadata(resultSet);
+            }
+
+            JSONArray facets;
+
+            try (PreparedStatement preparedStatement = connection.prepareStatement(
+                    Queries.getQuery("select_fingerprint_facet_relation"))) {
+                preparedStatement.setString(1, id);
+                preparedStatement.setString(2, instanceId);
+                ResultSet resultSet = preparedStatement.executeQuery();
+                facets = DataConversion.extractFacets(resultSet);
+            }
+
+            String json = DataConversion.constructFingerprintJSON(fingerprintMetadata, usageMetadata, facets);
```

Comment:
I would recommend creating one SQL statement that pulls the data you need for this fingerprintJson construction. You have three SQL query above, which is going to be more expensive than one. IMO you can easily work on a statement with a few joins to pull out all the data in one query :).

Reply:
Thanks @afalko for the review. I have changed it now to use a single query. Would love to hear your thoughts on it :)